File: android/app/src/main/java/com/returntracker/android/data/ReturnItem.kt
package com.returntracker.android.data

import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.Index
import androidx.room.PrimaryKey
import com.returntracker.android.domain.ReturnStatus
import java.time.LocalDate
import java.util.UUID

@Entity(
    tableName = "return_items",
    indices = [Index(value = ["sourceMessageId"], unique = true)],
)
data class ReturnItem(
    @PrimaryKey val id: String = UUID.randomUUID().toString(),
    val productName: String,
    val storeName: String,
    val priceWon: Long,
    val purchaseEpochDay: Long,
    val returnDeadlineEpochDay: Long,
    val orderNumber: String? = null,
    val note: String? = null,
    val statusRawValue: String = ReturnStatus.KEEPING.name,
    val createdAtMillis: Long = System.currentTimeMillis(),
    val updatedAtMillis: Long = System.currentTimeMillis(),
    val sourceRawValue: String = ReturnItemSource.MANUAL.name,
    val sourceMessageId: String? = null,
    val needsReview: Boolean = false,
) {
    @get:Ignore
    val purchaseDate: LocalDate
        get() = LocalDate.ofEpochDay(purchaseEpochDay)

    @get:Ignore
    val returnDeadline: LocalDate
        get() = LocalDate.ofEpochDay(returnDeadlineEpochDay)

    @get:Ignore
    val status: ReturnStatus
        get() = ReturnStatus.entries.firstOrNull { it.name == statusRawValue } ?: ReturnStatus.KEEPING

    @get:Ignore
    val source: ReturnItemSource
        get() = ReturnItemSource.entries.firstOrNull { it.name == sourceRawValue } ?: ReturnItemSource.MANUAL
}

enum class ReturnItemSource {
    MANUAL,
    GMAIL,
}

File: ReturnTracker/Models/ReturnItem.swift
import Foundation
import SwiftData

@Model
final class ReturnItem {
    @Attribute(.unique) var id: UUID
    var productName: String
    var storeName: String
    var price: Decimal
    var purchaseDate: Date
    var returnDeadline: Date
    var orderNumber: String?
    var note: String?
    var statusRawValue: String
    var createdAt: Date
    var updatedAt: Date
    var imageData: Data?
    var sourceRawValue: String = ReturnItemSource.manual.rawValue
    var sourceMessageID: String?
    var needsReview: Bool = false

    init(
        id: UUID = UUID(),
        productName: String,
        storeName: String,
        price: Decimal,
        purchaseDate: Date,
        returnDeadline: Date,
        orderNumber: String? = nil,
        note: String? = nil,
        status: ReturnStatus = .keeping,
        createdAt: Date = .now,
        updatedAt: Date = .now,
        imageData: Data? = nil,
        source: ReturnItemSource = .manual,
        sourceMessageID: String? = nil,
        needsReview: Bool = false
    ) {
        self.id = id
        self.productName = productName
        self.storeName = storeName
        self.price = price
        self.purchaseDate = purchaseDate
        self.returnDeadline = returnDeadline
        self.orderNumber = orderNumber
        self.note = note
        self.statusRawValue = status.rawValue
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.imageData = imageData
        self.sourceRawValue = source.rawValue
        self.sourceMessageID = sourceMessageID
        self.needsReview = needsReview
    }

    var status: ReturnStatus {
        get { ReturnStatus(rawValue: statusRawValue) ?? .keeping }
        set {
            statusRawValue = newValue.rawValue
            updatedAt = .now
        }
    }

    var source: ReturnItemSource {
        ReturnItemSource(rawValue: sourceRawValue) ?? .manual
    }
}
